skxray/core/correlation.py

```diff
@@ -367,3 +380,130 @@ def extract_label_indices(labels):
+    -------
+    g2 : array
+        normalized intensity-intensity time autocorreltion
+
+    Note :
+    The intensity-intensity autocorrelation g2 is connected to the intermediate
+    scattering factor(ISF) g1
+
+    :math ::
+        g2(q, t) = \\beta[g1(q, t)]^{2} + g_\\infty
+
+    For a system undergoing  diffusive dynamics,
+
+    :math ::
+        g1(q, t) = e^{-\\Gamma t}
+
+    :math ::
+        g2(q, t) = \\beta e^{-2\\Gamma t} + g_\\infty
+
+    These implementation are based on published work. [1]_
+
+     References
+    ----------
+
+    .. [1] L. Li, P. Kwasniewski, D. Orsi, L. Wiegart, L. Cristofolini, C. Caronna
+       and A. Fluerasu, " Photon statistics and speckle visibility spectroscopy with
+       partially coherent X-rays," J. Synchrotron Rad. vol 21, p 1288-1295, 2014
+
+    """
+    return beta*np.exp(-2*relaxation_rate*lags) + baseline
+
+
+def _residual_auto_corr(params, lags, g2_data, *eps_data):
+    """
+    This will provide difference between experiment data and the model
+    Parameters
+    ----------
+    params : dict or Parameters
+        beta - float, optical contrast (speckle contrast),
+        relaxation_rate - float, relaxation time associated with the
+        samples dynamics,
+        baseline -float, optional baseline of one time
+        correlation equal to one for ergodic samples
+        have to give as either dictionary or Parameters
+        eg: One of the following
+        #create a dictionary
+        {'beta': 0.2, 'relaxation_rate':10, 'baseline':1}
+
+        # create a set of Parameters
+        from lmfit import Parameters
+        params = Parameters()
+        params.add('beta',  value=0.2,  min=0, max=0.3)
+        params.add('relaxation_rate', value=10, min=0, max=11)
+        params.add('baseline', value=1)
+
+    lags : array
+        delay time
+
+    g2_data : array
+        normalized intensity-intensity time autocorreltion
+
+    eps_data : array, optional
+        standard error of the normalized intensity-intensity
+        time autocorreltion
+
+    Returns
+    -------
+    residual : array
+        difference between experimental result and the model
+    """
+    # create set of parameters
+    beta = params['beta'].value
+    relaxation_rate = params['relaxation_rate'].value
+    baseline = params['baseline'].value
+
+    return (g2_data - auto_corr_scat_factor(lags, beta, relaxation_rate,
+                                            baseline=1))/eps_data
+
+
+def fit_auto_corr(params, x, data, eps_data):
+    """
+    Minimize the function and calculate final result
+
+    Parameters
+    ----------
+    params: dict
+        parameters dictionary
+
+    x : array
+       x
+
+    data : array
+
+    eps_data : array
+
+    Returns
+    -------
+    final_result : array
+        minimized fitting to results using least square model
+
+    """
+    result = minimize(_residual_auto_corr, params, args=(x, data, eps_data))
+    return data + result.residual
```

Comment:
@tacaswell To get the final result after minimizing. Discussed this issue with @danielballan and removed that function because using lmfit Model class fitting can be done directly without another function.
(http://cars9.uchicago.edu/software/python/lmfit/model.html#model.Model)
example
-----------
from lmfit import Model
mod = Model(auto_corr_scat_factor)
fitted_result = mod.eval(lags=lags, beta=0.2234, relaxation_rate = 6.2567,  baseline=1.0)